plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/GcpLocalRunTab.java

```diff
@@ -0,0 +1,290 @@
+
+  private static final String ATTRIBUTE_ACCOUNT_EMAIL =
+      "com.google.cloud.tools.eclipse.gcpEmulation.accountEmail"; //$NON-NLS-1$
+
+  private static final String PROJECT_ID_ENVIRONMENT_VARIABLE =
+      "GOOGLE_CLOUD_PROJECT"; //$NON-NLS-1$
+  private static final String SERVICE_KEY_ENVIRONMENT_VARIABLE =
+      "GOOGLE_APPLICATION_CREDENTIALS"; //$NON-NLS-1$
+
+  private final EnvironmentTab environmentTab;
+  private final IGoogleLoginService loginService;
+  private final ProjectRepository projectRepository;
+
+  private AccountSelector accountSelector;
+  private ProjectSelector projectSelector;
+  private Text serviceKeyInput;
+
+  private Image gcpIcon;
+
+  // We set up intermediary models between a run configuration and UI components for certain values,
+  // because, e.g., the account selector cannot load an email if it is not logged in. In such a
+  // case, although nothing is selected in the account selector, we should not clear the email saved
+  // in the run configuration.
+  private String accountEmailModel;
+  private String gcpProjectIdModel;
+  // To prevent updating above models when programmatically setting up UI components.
+  private boolean initializingUiValues;
+
+  private boolean activated;
+
+  public GcpLocalRunTab(EnvironmentTab environmentTab) {
+    this(environmentTab, PlatformUI.getWorkbench().getService(IGoogleLoginService.class),
+        new ProjectRepository(PlatformUI.getWorkbench().getService(IGoogleApiFactory.class)));
+  }
+
+  @VisibleForTesting
+  GcpLocalRunTab(EnvironmentTab environmentTab,
+      IGoogleLoginService loginService, ProjectRepository projectRepository) {
+    this.environmentTab = environmentTab;
+    this.loginService = loginService;
+    this.projectRepository = projectRepository;
+  }
+
+  @Override
+  public void createControl(Composite parent) {
+    Composite composite = new Composite(parent, SWT.NONE);
+
+    // Account row
+    new Label(composite, SWT.LEAD).setText(Messages.getString("label.account")); //$NON-NLS-1$
+    accountSelector = new AccountSelector(composite, loginService);
+    accountSelector.addSelectionListener(new Runnable() {
+      @Override
+      public void run() {
+        updateProjectSelector();
+
+        if (!initializingUiValues) {
+          boolean emptySelection = accountSelector.getSelectedEmail().isEmpty();
+          boolean savedEmailAvailable = accountSelector.isEmailAvailable(accountEmailModel);
+          if (!emptySelection || savedEmailAvailable) {
+            accountEmailModel = accountSelector.getSelectedEmail();
+            gcpProjectIdModel = ""; //$NON-NLS-1$
+            updateLaunchConfigurationDialog();
+          }
+        }
+      }
+    });
+
+    // Project row
+    Label projectLabel = new Label(composite, SWT.LEAD);
+    projectLabel.setText(Messages.getString("label.project")); //$NON-NLS-1$
+
+    Composite projectSelectorComposite = new Composite(composite, SWT.NONE);
```

Comment:
I wonder if it makes sense to extract this into a `AccountProjectSelectionPanel` or something.  We use this in a couple of places, like the `AppEngineDeployDialog`.  And #1374 seems like it relevant here too — maybe we should turn `ProjectRepository` into an injectable service that caches the credential-projects for some period of time.

Reply:
Indeed, we may decide to do so in the future. For now, the deploy dialog uses databinding, and things work quite differently there (e.g., auto-selecting a single account, it has a refresh button, it has the App Engine application checking and error messages, it has the project creation link, etc), so the project selector there is very specialized. So, this would require a substantial work to factor out, so we'd probably want to consider it when the next time we need the bare minimum combination in this PR.

The cache idea is nice too.